### ISSUE TITLE
Feature: Coroutines

### ISSUE THREAD

**Issue by JSandusky** [JamesBoer/Jinx#11]
Coroutines/generator objects would be nice.

The issue in particular is calling functions that can themselves _wait_ while waiting on them as well as knowing that the functions completed to accomplish a race.
    
    -- this is a RACE, one coroutine will win the race
    wait until (A_that_waits or B_that_waits)
    if coro finished A_that_waits
        ... do stuff because A won the race
    else if coro finished B_that_waits
        ... do stuff because B won the race
    end

Or are there any workarounds that are close enough to a coroutine/generator?

**Comment by JamesBoer** [comment]
So, to help me understand a bit more clearly... When you say "this is a RACE, one coroutine will win the race", is that a statement of what WILL happen or what you WANT to happen?  And by "race" are you describing a data-race (a threading problem), or a literal race between co-routine execution.  That is, you want to detect which function will finish executing first?

**Comment by JSandusky** [comment]
Literal race. A race terminates when any coroutine participating finishes. Dual of a sync, which is waiting on all coroutines to finish. 

I've cludged it in by adding an `async` overload for function that tags the function as being an async function and that when it waits it pops the ExecutionFrame to store for future use then returns false. When called again, it just pulls the saved ExecutionFrame then either resumes or resets.

I have not yet cludged in testing coroutine status or reseting though. Given the natural language style of functions I can't figure out how to identify them in a fashion that isn't "_parse a damn string_" ... I'm probably going to force it in as "_parse a damn string, but do this one here at compile time_" like `` `{} silly function {}` `` or some such.

**Comment by JamesBoer** [comment]
Got it.  I'll have to give this some thought to come up with a possible design.  No promises, but will definitely investigate feasibility.

**Comment by JamesBoer** [comment]
I've got an initial design I'm gathering feedback on here:
https://www.reddit.com/r/ProgrammingLanguages/comments/k8xpfa/advice_desired_adding_coroutines_to_jinx/

**Comment by JamesBoer** [comment]
Quick status update: I've done a bit of prototyping work on this, and ran into a few snags in the way script objects maintain stored variables in the execution stack, making things a bit messier than I had hoped.  I think I'm going to take a step back and see if I can implement first-class functions as a feature, then build on top of that, with a few changes to the script execution stack to better support coroutines afterwards.

**Comment by JamesBoer** [comment]
Experimental coroutine support can be found in branch **feature-coroutines**.  The new features are fully documented in the tutorial and library API docs.  I'll just show you a quick example of what it looks like:
```
import core

-- Function declaration
function count to {integer y}
    set x to 0
    loop while x < y
        increment x
        wait
    end
    return x
end

-- Execute functions asynchronously and store coroutine in variables
set c1 to async call function count to {} with 15
set c2 to async call function count to {} with 10
set c3 to async call function count to {} with 5

-- Wait until any coroutine is finished
wait until any of c1, c2, c3 is finished

-- Retrieve return value from coroutine
set v to c1's value -- c1 = null
set v to c2's value -- c2 = null
set v to c3's value -- c3 = 5
```

Let me know if you have any suggestions or run into any issues.

**Comment by JamesBoer** [comment]
First-class function and coroutine support have been merged into development branch.

**Comment by JamesBoer** [comment]
Feature has been merged into master branch.

**Comment by JamesBoer** [comment]
Sure thing.  Let me know if you run into any issues after you get around to trying it out.
